src/github.com/matrix-org/dendrite/appservice/workers/transaction_scheduler.go

```diff
@@ -0,0 +1,226 @@
+	workerStates []types.ApplicationServiceWorkerState,
+) error {
+	// Create a worker that handles transmitting events to a single homeserver
+	for _, workerState := range workerStates {
+		// Don't create a worker if this AS doesn't want to receive events
+		if workerState.AppService.URL != "" {
+			go worker(appserviceDB, workerState)
+		}
+	}
+	return nil
+}
+
+// worker is a goroutine that sends any queued events to the application service
+// it is given.
+func worker(db *storage.Database, ws types.ApplicationServiceWorkerState) {
+	ctx := context.Background()
+
+	// Initialize transaction ID counter
+	var err error
+	currentTransactionID, err = db.GetTxnIDWithAppServiceID(ctx, ws.AppService.ID)
+	if err != nil && err != sql.ErrNoRows {
+		log.WithError(err).Fatalf("appservice %s worker unable to get latest transaction ID from DB",
+			ws.AppService.ID)
+		return
+	}
+
+	// Grab the HTTP client for sending requests to app services
+	client := &http.Client{
+		Timeout: transactionTimeout,
+	}
+
+	// Initial check for any leftover events to send from last time
+	eventCount, err := db.CountEventsWithAppServiceID(ctx, ws.AppService.ID)
+	if err != nil {
+		log.WithError(err).Fatalf("appservice %s worker unable to read queued events from DB",
+			ws.AppService.ID)
+		return
+	}
+
+	// Wait if there are no new events to go out
+	if eventCount == 0 {
+		waitForEvents(&ws)
+	}
+
+	// Loop forever and keep waiting for more events to send
+	for {
+		// Set EventsReady to false for some reason (we just sent events?)
+		ws.Cond.L.Lock()
+		ws.EventsReady = false
+		ws.Cond.L.Unlock()
+
+		eventIDs, events, err := db.GetEventsWithAppServiceID(ctx, ws.AppService.ID, transactionBatchSize)
+		if err != nil {
+			log.WithError(err).Errorf("appservice %s worker unable to read queued events from DB",
+				ws.AppService.ID)
+
+			// Wait a little bit for DB to possibly recover
+			time.Sleep(transactionBreakTime)
+			continue
+		}
+
+		// Batch events up into a transaction
+		transactionJSON, err := createTransaction(events)
+		if err != nil {
+			log.WithError(err).Fatalf("appservice %s worker unable to marshal events",
+				ws.AppService.ID)
+
+			return
+		}
+
+		// Send the events off to the application service
+		err = send(client, ws.AppService, transactionJSON)
+		if err != nil {
+			// Backoff
+			backoff(err, &ws)
+			continue
+		}
+
+		// We sent successfully, hooray!
+		ws.Backoff = 0
+
+		// Remove sent events from the DB
+		err = db.RemoveEventsBeforeAndIncludingID(ctx, eventIDs[len(eventIDs)-1])
+		if err != nil {
+			log.WithError(err).Fatalf("unable to remove appservice events from the database for %s",
+				ws.AppService.ID)
+			return
+		}
+
+		// Update transactionID
+		currentTransactionID++
+		if err = db.UpsertTxnIDWithAppServiceID(ctx, ws.AppService.ID, currentTransactionID); err != nil {
+			log.WithError(err).Fatalf("unable to update transaction ID for %s",
+				ws.AppService.ID)
+			return
+		}
+		waitForEvents(&ws)
```

Comment:
What happens if we hit `transactionBatchSize` in the transaction in [104](https://github.com/matrix-org/dendrite/pull/477/files#diff-d84306218ea8217c1024f21eb34d9fcbR104)
We should not wait for the next Broadcast until we process further queries when we have pending events in the database

Reply:
Good point, the earlier implementation dealt with this, but as we no longer use an event counter, this did not.

I've added to the query a count for the total events, and we won't sleep until that is 0. Adding this count is slightly more expensive that before, but everything is still done in one query, so it should still be fast.